cur-lib/cur/curnel/turnstile-impl/dep-ind-cur2.rkt

```diff
@@ -90,72 +97,89 @@
-               (typecheck? (subst #'x2 #'x1 #'τ_out1) #'τ_out2))]
-         [_ #;[(printf "failed to type check: ~a\n" (syntax->datum this-syntax))] #f])))))
-
-;; lambda and #%app -----------------------------------------------------------
-(define-typed-syntax λ/1
-  ;; expected ty only
+     ; Since universes are type checked lazily, if the type we're checking (t1)
+     ; is a universe, force it,
+     (define t1+
+       (syntax-parse t1
+         [((~literal Type) _) ((current-type-eval) t1)]
+         [_ t1]))
+     (or (type=? t1+ t2) ; equality
+         (syntax-parse (list t1+ t2)
+           [((~Type n) (~Type m)) (<= (stx-e #'n) (stx-e #'m))]
+           [((~Π [x1 : τ_in1] τ_out1) (~Π [x2 : τ_in2] τ_out2))
+            (and (type=? #'τ_in1 #'τ_in2)
+                 (typecheck? (subst #'x2 #'x1 #'τ_out1) #'τ_out2))]
+           ; NB: Overwrite, do not preserve, old relation.
+           [_ #f])))))
+
+;;; Term Definitions
+;;; -----------------------------------------------------------
+
+;; Function and application
+;; ------------------------------------------------
+(define-typed-syntax λ
+  ; expected ty only
   [(_ y:id e) ⇐ (~Π [x:id : τ_in] τ_out) ≫
    [[x ≫ x- : τ_in] ⊢ #,(subst #'x #'y #'e) ≫ e- ⇐ τ_out]
    ---------
    [⊢ (λ- (x-) e-) ⇒ (Π [x : τ_in] τ_out)]]
-  ;; both expected ty and annotations
+
+  ; both expected ty and annotations
   [(_ [y:id (~datum :) τ_in*] e) ⇐ (~Π [x:id : τ_in] τ_out) ≫
    [⊢ τ_in* ≫ τ_in** ⇒ (~Type _)]
    #:when (typecheck? #'τ_in** #'τ_in)
    [[x ≫ x- : τ_in] ⊢ [#,(subst #'x #'y #'e) ≫ e- ⇐ τ_out]]
    -------
    [⊢ (λ- (x-) e-) ⇒ (Π [x : τ_in] τ_out)]]
-  ;; annotations only
+
+  ; annotations only
   [(_ [x:id (~datum :) τ_in] e) ≫
-   [[x ≫ x- : τ_in] ⊢ [e ≫ e- ⇒ τ_out] [τ_in ≫ τ_in- ⇒ _]]
+   [[x ≫ x-- : τ_in] ⊢ [e ≫ e- ⇒ τ_out] [τ_in ≫ τ_in- ⇒ _]]
+   #:with x-
+   (transfer-prop 'display-as #'x (transfer-prop 'tmp #'x #'x--))
    -------
-   [⊢ (λ- (x-) e-) ⇒ #,(quasisyntax/loc this-syntax (Π [#,(transfer-prop 'display-as #'x (transfer-prop 'tmp #'x #'x-)) : τ_in-] τ_out))]])
+   [⊢ (λ- (x-) e-) ⇒ (Π [x- : τ_in-] τ_out)]])
+
+(begin-for-syntax
+  ; Pattern for Racket's #%plain-lambda
+  (define-syntax ~λ-
+    (pattern-expander
+     (syntax-parser
+       [(_ (x:id ...) e ...)
+        #`(~or ((~literal #%plain-lambda) (x ...) e ...)
+               ; TODO: Who adds #%expression?
+               ((~literal #%expression) ((~literal #%plain-lambda) (x ...) e ...)))]))))
 
 (define-typerule/red (app e_fn e_arg) ≫
   [⊢ e_fn ≫ e_fn- ⇒ (~Π [X : τ_in] τ_out)]
   [⊢ e_arg ≫ e_arg- ⇐ τ_in]
   #:with τ_out- (reflect (subst #'e_arg- #'X #'τ_out)) ; TODO: fix orig
   -----------------------------
-  [⊢ (app/eval e_fn- e_arg-) ⇒ τ_out- #;(app/eval (λ- (X) τ_out) e_arg-)]
+  [⊢ (app/eval e_fn- e_arg-) ⇒ τ_out-]
   #:where app/eval
-  [(#%plain-app
-    (~or ((~literal #%plain-lambda) (x) e)
-         ((~literal #%expression) ((~literal #%plain-lambda) (x) e))) ; TODO: who adds this?
```

Comment:
Any chance you tried removing the `#%expression` case? If not I'll experiment with it. It might be vestigial code from an earlier version of `define-red`

Reply:
I didn't try it.